# EDA - Pokemon Legendary Classifier

In [1]:
# Import libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns

In [2]:
# Reading our clean pokemon dataset
df_pokemon = pd.read_csv(r"..\..\Data_Scientist_Training\Pokemon-Legendary-Classifier\datasets\pokemon_cleaned.csv")

In [3]:
# Checking out our dataset
df_pokemon.head(3)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
0,1,Bulbasaur,grass,poison,45,49,49,65,65,45,318,0
1,2,Ivysaur,grass,poison,60,62,63,80,80,60,405,0
2,3,Venusaur,grass,poison,80,100,123,122,120,80,625,0


In [4]:
# Checking out the shape of our dataset
df_pokemon.shape

(797, 12)

## Perform Exploring Data - EDA Technique

There are two common techniques to perform EDA: exploring and visualizing. In the exploration phase, we filter, group, and sort our data to summarize it and build a clearer understanding of what we are working with.

Keep in mind that the goal of EDA is to examine our data from multiple angles; uncovering patterns, relationships, and insights along the way. For example, we can explore how our data is distributed between normal and legendary Pokémon

In [5]:
# Getting an statistical overview from our dataset
df_pokemon.describe()

,pokedex_number,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
count,797.000000,797.000000,797.000000,797.000000,797.000000,797.000000,797.000000,797.000000,797.000000
mean,401.943538,68.890841,77.721455,72.841907,71.431619,70.953576,66.360100,428.199498,0.087829
std,231.523934,26.594042,32.078151,30.684856,32.376040,27.949553,28.910268,119.194824,0.283224
min,1.000000,1.000000,5.000000,5.000000,10.000000,20.000000,5.000000,180.000000,0.000000
25%,201.000000,50.000000,55.000000,50.000000,46.000000,50.000000,45.000000,320.000000,0.000000
50%,403.000000,65.000000,75.000000,70.000000,65.000000,66.000000,65.000000,435.000000,0.000000
75%,602.000000,80.000000,100.000000,90.000000,92.000000,90.000000,85.000000,505.000000,0.000000
max,801.000000,255.000000,185.000000,230.000000,194.000000,230.000000,180.000000,780.000000,1.000000


### Filtering Pokemon by stats, type and legendary record

Legendary Pokémon tend to have higher stats, so we will filter Pokémon with a total stat sum greater than 500 and check which ones are legendary.

Additionally, we'll review the types for each Pokémon. Some legendary Pokémon tend to have Psychic or Dragon as their primary type, and often feature dual typing.

In [6]:
# Filtering Pokemon whose stats are greater than 500
stats_mask = df_pokemon['Total Stats'] > 500

filtered_pokemon = df_pokemon[stats_mask]

In [7]:
# Checking out legendary pokemons and its stats
filtered_pokemon[filtered_pokemon['is_legendary'] == 1]

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
142,144,Articuno,ice,flying,90,85,100,95,125,85,580,1
143,145,Zapdos,electric,flying,90,90,85,125,90,100,580,1
144,146,Moltres,fire,flying,90,100,90,125,85,90,580,1
148,150,Mewtwo,psychic,NaN,106,150,70,194,120,140,780,1
149,151,Mew,psychic,NaN,100,100,100,100,100,100,600,1
...,...,...,...,...,...,...,...,...,...,...,...,...
792,797,Celesteela,steel,flying,97,101,103,107,101,61,570,1
793,798,Kartana,grass,steel,59,181,131,59,31,109,570,1
794,799,Guzzlord,dark,dragon,223,101,53,97,53,43,570,1
795,800,Necrozma,psychic,NaN,97,107,101,127,89,79,600,1


In [8]:
# Filter by pokemon whose second type is null
filtered_pokemon[filtered_pokemon['type2'].isna()]

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
8,9,Blastoise,water,NaN,79,103,120,135,115,78,630,0
58,59,Arcanine,fire,NaN,90,110,80,100,80,95,555,0
64,65,Alakazam,psychic,NaN,55,50,65,175,105,150,600,0
67,68,Machamp,fighting,NaN,90,130,80,65,85,55,505,0
113,115,Kangaskhan,normal,NaN,105,125,100,60,100,100,590,0
...,...,...,...,...,...,...,...,...,...,...,...,...
758,763,Tsareena,grass,NaN,72,120,98,50,98,72,510,0
767,772,Type: Null,normal,NaN,95,95,95,95,95,59,534,0
768,773,Silvally,normal,NaN,95,95,95,95,95,95,570,0
791,796,Xurkitree,electric,NaN,83,89,71,173,71,83,570,1


In [9]:
type1_mask = (filtered_pokemon['type1'] == 'psychic') | (filtered_pokemon['type1'] == 'dragon')

filtered_pokemon[type1_mask]

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
64,65,Alakazam,psychic,NaN,55,50,65,175,105,150,600,0
147,149,Dragonite,dragon,flying,91,134,95,100,100,80,600,0
148,150,Mewtwo,psychic,NaN,106,150,70,194,120,140,780,1
149,151,Mew,psychic,NaN,100,100,100,100,100,100,600,1
194,196,Espeon,psychic,NaN,65,65,60,130,95,110,525,0
245,249,Lugia,psychic,flying,106,90,130,90,154,110,680,1
247,251,Celebi,psychic,grass,100,100,100,100,100,100,600,1
278,282,Gardevoir,psychic,fairy,68,85,65,165,135,100,618,0
329,334,Altaria,dragon,flying,75,110,110,110,105,80,590,0
368,373,Salamence,dragon,flying,95,145,130,120,90,120,700,0


Based on the filtering operations performed previously, we found that not all legendary Pokémon have a total stat sum greater than 500. Some of the filtered Pokémon are actually final evolutions, such as Blastoise and Alakazam.

Additionally, there is a total of 68 legendary Pokémon, which we'll need to distribute across our sets for ML training. We also confirmed that most legendaries have Psychic or Dragon as their primary type, a useful insight that will support our classifier.

### Sorting Pokemon by each stat

Now we'll sort our data based on Pokémon stats to get an overall view of which Pokémon have the highest or lowest values in a specific stat. For example, legendary Pokémon tend to have lower Speed but higher HP compared to regular ones.

In [10]:
# Sorting pokemon by Total Stats on descending order
filtered_pokemon.sort_values('Total Stats', ascending = False)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
148,150,Mewtwo,psychic,NaN,106,150,70,194,120,140,780,1
379,384,Rayquaza,dragon,flying,105,180,100,180,100,115,780,1
378,383,Groudon,ground,NaN,100,180,160,150,90,90,770,1
377,382,Kyogre,water,NaN,100,150,90,180,160,90,770,1
488,493,Arceus,normal,NaN,120,120,120,120,120,120,720,1
...,...,...,...,...,...,...,...,...,...,...,...,...
67,68,Machamp,fighting,NaN,90,130,80,65,85,55,505,0
33,34,Nidoking,poison,ground,81,102,77,85,75,85,505,0
425,430,Honchkrow,dark,flying,100,125,52,105,52,71,505,0
529,534,Conkeldurr,fighting,NaN,105,140,95,55,65,45,505,0


In [11]:
# Sorting HP in ascending order
filtered_pokemon.sort_values('hp', ascending = True)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
210,213,Shuckle,bug,rock,20,10,230,10,230,5,505,0
472,477,Dusknoir,ghost,NaN,45,100,135,65,135,45,525,0
741,746,Wishiwashi,water,NaN,45,140,130,140,135,30,620,0
381,386,Deoxys,psychic,NaN,50,95,90,95,90,180,600,1
474,479,Rotom,electric,ghost,50,65,107,105,107,86,520,0
...,...,...,...,...,...,...,...,...,...,...,...,...
284,289,Slaking,normal,NaN,150,160,100,95,65,100,670,0
141,143,Snorlax,normal,NaN,160,110,65,65,110,30,540,0
713,718,Zygarde,dragon,ground,216,100,121,91,95,85,708,1
794,799,Guzzlord,dark,dragon,223,101,53,97,53,43,570,1


In [12]:
# Sorting by speed
filtered_pokemon.sort_values('speed', ascending = True)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
210,213,Shuckle,bug,rock,20,10,230,10,230,5,505,0
318,323,Camerupt,fire,ground,70,120,100,145,105,20,560,0
521,526,Gigalith,rock,NaN,85,135,130,60,80,25,515,0
708,713,Avalugg,ice,NaN,95,117,184,44,46,28,514,0
206,208,Steelix,steel,ground,75,125,230,55,95,30,610,0
...,...,...,...,...,...,...,...,...,...,...,...,...
250,254,Sceptile,grass,NaN,70,110,75,145,85,145,630,0
140,142,Aerodactyl,rock,flying,80,135,85,70,95,150,615,0
64,65,Alakazam,psychic,NaN,55,50,65,175,105,150,600,0
790,795,Pheromosa,bug,fighting,71,137,37,137,37,151,570,1


In [13]:
# Sorting by attack
filtered_pokemon.sort_values('attack', ascending = False)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
211,214,Heracross,bug,fighting,80,185,115,40,105,75,600,0
793,798,Kartana,grass,steel,59,181,131,59,31,109,570,1
378,383,Groudon,ground,NaN,100,180,160,150,90,90,770,1
379,384,Rayquaza,dragon,flying,105,180,100,180,100,115,780,1
440,445,Garchomp,dragon,ground,108,170,115,120,95,92,700,0
...,...,...,...,...,...,...,...,...,...,...,...,...
64,65,Alakazam,psychic,NaN,55,50,65,175,105,150,600,0
610,615,Cryogonal,ice,NaN,80,50,50,95,135,105,515,0
550,555,Darmanitan,fire,fire,105,30,105,140,105,55,540,0
210,213,Shuckle,bug,rock,20,10,230,10,230,5,505,0


In [14]:
filtered_pokemon.sort_values('sp_attack', ascending = False)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
148,150,Mewtwo,psychic,NaN,106,150,70,194,120,140,780,1
377,382,Kyogre,water,NaN,100,150,90,180,160,90,770,1
379,384,Rayquaza,dragon,flying,105,180,100,180,100,115,780,1
64,65,Alakazam,psychic,NaN,55,50,65,175,105,150,600,0
791,796,Xurkitree,electric,NaN,83,89,71,173,71,83,570,1
...,...,...,...,...,...,...,...,...,...,...,...,...
467,472,Gliscor,ground,flying,75,95,125,45,75,95,510,0
456,461,Weavile,dark,ice,70,120,65,45,85,125,510,0
708,713,Avalugg,ice,NaN,95,117,184,44,46,28,514,0
211,214,Heracross,bug,fighting,80,185,115,40,105,75,600,0


As we can see from the previous sorting operations, legendary Pokémon tend to have a higher combined stat total. Regarding each specific stat, there is significant variation. For example:

- Speed: Pokémon at their final stage of evolution (Sceptile or Alakazam) have similar Speed stats to Pheromosa, which is considered one of the fastest Ultra Beasts.
- HP: Pokémon with no evolution line or at their final evolution stage dominate this stat. Blissey, Snorlax, and Slaking are not legendary Pokémon yet have the highest HP concentration — showing a high diversity in this field.
- Sp. Attack: Most legendaries have higher Special Attack, with only a few final evolutions ranking at the top.

Based on these insights, the Total Stats column could be a strong feature for our classifier.

**Note**: For our Legendary Classifier, we will consider **Ultra Beasts** as Legendary Pokémon.

### Grouping Data by Types

To better understand how legendary Pokémon are distributed by type, we'll group them by the most common types among legendaries. Which, based on our filtering step, are Dragon and Psychic.

In [15]:
# Getting a head-count of each pokemon type
filtered_pokemon.groupby(['type1'])['pokedex_number'].count().sort_values(ascending = False)

type1
water       26
fire        19
psychic     19
grass       18
normal      16
dragon      14
electric    13
dark        11
steel       11
rock        11
ice         10
bug          9
ground       8
ghost        6
fairy        5
fighting     5
poison       3
flying       2
Name: pokedex_number, dtype: int64

In [16]:
# Creating mask for legendary pokemons
legendary_mask = df_pokemon['is_legendary'] == 1

# Filtering our data to get only legendary pokemons
df_legendary_pkmn = df_pokemon[legendary_mask]

In [17]:
# Checking out our legendary dataframe
df_legendary_pkmn.head(3)

,pokedex_number,name,type1,type2,hp,attack,defense,sp_attack,sp_defense,speed,Total Stats,is_legendary
142,144,Articuno,ice,flying,90,85,100,95,125,85,580,1
143,145,Zapdos,electric,flying,90,90,85,125,90,100,580,1
144,146,Moltres,fire,flying,90,100,90,125,85,90,580,1


In [18]:
df_legendary_pkmn.groupby(['type1'])['pokedex_number'].count().sort_values(ascending = False)

type1
psychic     17
dragon       7
steel        6
water        6
fire         5
electric     5
grass        4
rock         4
dark         3
normal       3
bug          3
ice          2
ground       2
ghost        1
fairy        1
flying       1
Name: pokedex_number, dtype: int64

As we can see, Dragon and Psychic are the most common types among legendary Pokémon, confirming the hypothesis we formed during the filtering section.

In [19]:
# Checking out if stats ranges overlaps
df_legendary_pkmn.groupby(['is_legendary'])['Total Stats'].agg(['min', 'max', 'std']).reset_index()

,is_legendary,min,max,std
0,1,200,780,82.63499


Finally, we perform some aggregation functions to get a statistical overview of legendary Pokémon with respect to their Total Stats. Some key insights are:

The Standard Deviation is `82.63`, meaning most of the data is distributed within ~80 points of the mean.
The minimum value is 200 and the maximum value is 780. While the maximum seems reasonable, the minimum value appears to be an outlier.

With this in mind, let's find the Pokémon whose Total Stats equal 200.

In [20]:
df_legendary_pkmn['Total Stats'].sort_values(ascending = True)

784    200
785    400
788    570
783    570
780    570
      ... 
488    720
378    770
377    770
148    780
379    780
Name: Total Stats, Length: 70, dtype: int64

In [21]:
# Who is the pokemon with lowest Total Stats
df_legendary_pkmn.loc[784]

pokedex_number        789
name               Cosmog
type1             psychic
type2                 NaN
hp                     43
attack                 29
defense                31
sp_attack              29
sp_defense             31
speed                  37
Total Stats           200
is_legendary            1
Name: 784, dtype: object

The Pokémon with the lowest stats turns out to be Cosmog. To confirm whether this represents an outlier, we'll move into the Visualization phase to detect any anomalies that could hurt the performance of our classifier.

With this, we complete the Exploration phase of our EDA process. Next, we'll move into the Visualization phase to identify patterns, trends, and potential anomalies.